src/webgpu/shader/validation/expression/call/builtin/textureSample.spec.ts

```diff
@@ -0,0 +1,254 @@
+* test textureSample coords parameter must be correct type
+* test textureSample offset parameter must be correct type
+* test textureSample offset parameter must be a const-expression
+* test textureSample offset parameter must be between -8 and +7 inclusive
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import {
+  kConcreteIntegerScalarsAndVectors,
+  kConvertableToFloatScalarsAndVectors,
+} from '../../../../../util/conversion.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+type TextureSampleParams = {
+  coords: string;
+  array_index?: boolean;
+  offset?: string;
+};
+
+const kValidTextureSampleParameterTypes: { [n: string]: TextureSampleParams } = {
+  'texture_1d<f32>': { coords: 'f32' },
+  'texture_2d<f32>': { coords: 'vec2<f32>', offset: 'vec2<i32>' },
+  'texture_2d_array<f32>': { coords: 'vec2<f32>', array_index: true, offset: 'vec2<i32>' },
+  'texture_3d<f32>': { coords: 'vec3<f32>', offset: 'vec3<i32>' },
+  'texture_cube<f32>': { coords: 'vec3<f32>' },
+  'texture_cube_array<f32>': { coords: 'vec3<f32>', array_index: true },
+  texture_depth_2d: { coords: 'vec2<f32>', offset: 'vec2<i32>' },
+  texture_depth_2d_array: { coords: 'vec2<f32>', array_index: true },
+  texture_depth_cube: { coords: 'vec3<f32>' },
+  texture_depth_cube_array: { coords: 'vec3<f32>', array_index: true },
+} as const;
+
+const kTextureTypes = keysOf(kValidTextureSampleParameterTypes);
+const kValuesTypes = objectsToRecord([
+  ...kConvertableToFloatScalarsAndVectors,
+  ...kConcreteIntegerScalarsAndVectors,
+] as const);
+
+function innerType(type: string) {
+  const angleNdx = type.indexOf('<');
+  return type.substring(angleNdx + 1, type.length - (angleNdx >= 0 ? 1 : 0));
+}
+
+// Replace with non-hacky version
+function isConvertible(src: string, dst: string) {
+  if (src === dst) {
+    return true;
+  }
+  const angleNdx = src.indexOf('<');
+  if (src.substring(0, angleNdx) !== dst.substring(0, angleNdx)) {
+    return false;
+  }
+
+  src = src.substring(angleNdx + 1, src.length - (angleNdx >= 0 ? 1 : 0));
+  dst = dst.substring(angleNdx + 1, dst.length - (angleNdx >= 0 ? 1 : 0));
+
+  switch (src) {
+    case 'abstract-float':
+      switch (dst) {
+        case 'abstract-float':
+        case 'f16':
+        case 'f32':
+        case 'f64':
+          return true;
+        default:
+          return false;
+      }
+    case 'abstract-int':
+      switch (dst) {
+        case 'abstract-int':
+        case 'abstract-float':
+        case 'f16':
+        case 'f32':
+        case 'f64':
+        case 'u16':
+        case 'u32':
+        case 'u8':
+        case 'i16':
+        case 'i32':
+        case 'i8':
+          return true;
+        default:
+          return false;
+      }
+    default:
+      return false;
+  }
+}
```

Comment:
```suggestion
/** @returns true if an argument of type 'src' can be used for a parameter of type 'dst' */
function isConvertible(src: Type, dst: Type) {
  if (src === dst) {
    return true;
  }

  const widthOf = (ty: Type) => {
    return ty instanceof VectorType ? ty.width : 1;
  };

  if (widthOf(src) !== widthOf(dst)) {
    return false;
  }

  const elSrc = scalarTypeOf(src);
  const elDst = scalarTypeOf(dst);

  switch (elSrc.kind) {
    case 'abstract-float':
      switch (elDst.kind) {
        case 'abstract-float':
        case 'f16':
        case 'f32':
        case 'f64':
          return true;
        default:
          return false;
      }
    case 'abstract-int':
      switch (elDst.kind) {
        case 'abstract-int':
        case 'abstract-float':
        case 'f16':
        case 'f32':
        case 'f64':
        case 'u16':
        case 'u32':
        case 'u8':
        case 'i16':
        case 'i32':
        case 'i8':
          return true;
        default:
          return false;
      }
    default:
      return false;
  }
}
```

Might be useful to promote this up to someplace reusable

Reply:
Note that `'f64'`, `'u16'`, `'u8'`, `'i16'`, `'i32'`, `'i8'` do not exist in WGSL. I'm guessing they were added to the framework for future compatibility, but you don't really need to handle them here.